[chore] Refactor `info` schema

kgrhartlage opened the pull request:
### Description

We previously referenced the root `info` schema for the `steps.<memberName>.info` schema definition. This adds needless complexity for including/excluding the `uuid` and `version` properties and also encourages the usage of `info` properties that take no effect at the step level (`author`, `slug`, `tags`).

As a bonus win, the docs struggle to handle nested `allOf` logic and this PR removes one such instance.

> **Note** I want to rename and sort files before merging!

kgrhartlage commented on src/schemata/definitions/step_template.yml:
```diff
@@ -1,22 +1,27 @@
+              The display name of the step in Laboperator.
+            examples:
+              - Tare Balance
+          description:
+            type: string
+            description: |
+              A descriptive summary of the step.
```
While this doesn't serve an explicit purpose, I suppose not everyone is aware of `$comment` and might want to use a description instead.

kgrhartlage commented on src/schemata/definitions/step_template.yml:
```diff
@@ -1,22 +1,27 @@
-        $ref: '#/definitions/info'
+        type: object
+        required: [title]
+        additionalProperties: false
```
The addition of `additionalProperties` and consequently will no longer allow things like `steps.<memberName>.info.author`. See the respective changes to `workflow_template_example` to fix the spec.

I see value in letting template authors know that any additional configuration here is ignored by Laboperator and should thus not be included. Can we get away with introducing this change?

ignatiusreza replied:
This is fine, it won't break existing uploaded templates..

kgrhartlage commented on src/schemata/definitions/schemaVersion.yml:
```diff
@@ -1,6 +1,7 @@
 ---
 title: Schema Version
-description: A valid Workflow schema version.
+description: |
+  A valid Workflow Schema version.
```
Previously this would be a `span` element, not it's using our mapped custom Typography component like all other descriptions.

kgrhartlage commented on src/schemata/definitions/info.yml:
```diff
@@ -1,33 +1,33 @@
+    $ref: '#/definitions/version'
   description:
     type: string
-    title: Description
```
<img width="537" alt="CleanShot 2023-04-12 at 10 06 05@2x" src="https://user-images.githubusercontent.com/42832087/231393552-1776aeb4-3cda-4251-9a67-56ab82382cf5.png">


[insert spinder-man pointing meme]

ignatiusreza replied:
![pointer](https://media.wdwnt.com/2023/04/903819-featured.jpg)

kgrhartlage commented on src/schemata/definitions/info.yml:
```diff
@@ -1,33 +1,35 @@
     type: string
-    title: Description
-    default: ''
+    maxLength: 10000
```
This field was added.

<details>
<summary>10000 chars</summary>

Lorem ipsum dolor sit amet consectetur adipisicing elit. Porro libero inventore voluptatibus ullam temporibus possimus ad odio fuga tempora atque qui consectetur eum voluptates numquam, commodi repellendus, architecto dolorum officiis nihil exercitationem, quidem impedit. Accusantium iusto vel placeat! Nihil ipsa aliquam maiores minus eum quos aperiam, minima aut nulla ipsam cumque, ex impedit assumenda velit ab tempore distinctio. Magnam repellendus doloribus suscipit itaque. Assumenda officia sed blanditiis modi iure quos voluptatibus labore? Dolorem nisi obcaecati adipisci officia eos porro laboriosam beatae ipsum molestiae eligendi recusandae libero sunt expedita veritatis dicta saepe vel, aliquam esse dolores voluptate, rerum, in enim mollitia deserunt? Ipsa omnis magni in eligendi vel facilis soluta vero officiis reprehenderit dolor eaque, cumque explicabo quo aliquam voluptatem distinctio vitae sit nam optio maxime nulla facere iste commodi hic. Quos repellendus sapiente accusamus quaerat, cumque facere perferendis quo iusto magnam omnis eveniet! Est sapiente laudantium quia, cum consectetur quo reiciendis assumenda, sunt, inventore quibusdam iusto dolores. Veritatis dolores blanditiis iusto omnis fugit cum magnam consectetur perspiciatis dolorem corporis sint sit ex esse nostrum, deserunt dolor delectus a voluptates amet eaque. Reprehenderit nam nemo esse provident accusamus nulla fuga dolorum aspernatur voluptatem, earum, explicabo repudiandae pariatur veniam, consequuntur a qui?

Lorem ipsum dolor sit amet consectetur adipisicing elit. Porro libero inventore voluptatibus ullam temporibus possimus ad odio fuga tempora atque qui consectetur eum voluptates numquam, commodi repellendus, architecto dolorum officiis nihil exercitationem, quidem impedit. Accusantium iusto vel placeat! Nihil ipsa aliquam maiores minus eum quos aperiam, minima aut nulla ipsam cumque, ex impedit assumenda velit ab tempore distinctio. Magnam repellendus doloribus suscipit itaque. Assumenda officia sed blanditiis modi iure quos voluptatibus labore? Dolorem nisi obcaecati adipisci officia eos porro laboriosam beatae ipsum molestiae eligendi recusandae libero sunt expedita veritatis dicta saepe vel, aliquam esse dolores voluptate, rerum, in enim mollitia deserunt? Ipsa omnis magni in eligendi vel facilis soluta vero officiis reprehenderit dolor eaque, cumque explicabo quo aliquam voluptatem distinctio vitae sit nam optio maxime nulla facere iste commodi hic. Quos repellendus sapiente accusamus quaerat, cumque facere perferendis quo iusto magnam omnis eveniet! Est sapiente laudantium quia, cum consectetur quo reiciendis assumenda, sunt, inventore quibusdam iusto dolores. Veritatis dolores blanditiis iusto omnis fugit cum magnam consectetur perspiciatis dolorem corporis sint sit ex esse nostrum, deserunt dolor delectus a voluptates amet eaque. Reprehenderit nam nemo esse provident accusamus nulla fuga dolorum aspernatur voluptatem, earum, explicabo repudiandae pariatur veniam, consequuntur a qui?

Lorem ipsum dolor sit amet consectetur adipisicing elit. Porro libero inventore voluptatibus ullam temporibus possimus ad odio fuga tempora atque qui consectetur eum voluptates numquam, commodi repellendus, architecto dolorum officiis nihil exercitationem, quidem impedit. Accusantium iusto vel placeat! Nihil ipsa aliquam maiores minus eum quos aperiam, minima aut nulla ipsam cumque, ex impedit assumenda velit ab tempore distinctio. Magnam repellendus doloribus suscipit itaque. Assumenda officia sed blanditiis modi iure quos voluptatibus labore? Dolorem nisi obcaecati adipisci officia eos porro laboriosam beatae ipsum molestiae eligendi recusandae libero sunt expedita veritatis dicta saepe vel, aliquam esse dolores voluptate, rerum, in enim mollitia deserunt? Ipsa omnis magni in eligendi vel facilis soluta vero officiis reprehenderit dolor eaque, cumque explicabo quo aliquam voluptatem distinctio vitae sit nam optio maxime nulla facere iste commodi hic. Quos repellendus sapiente accusamus quaerat, cumque facere perferendis quo iusto magnam omnis eveniet! Est sapiente laudantium quia, cum consectetur quo reiciendis assumenda, sunt, inventore quibusdam iusto dolores. Veritatis dolores blanditiis iusto omnis fugit cum magnam consectetur perspiciatis dolorem corporis sint sit ex esse nostrum, deserunt dolor delectus a voluptates amet eaque. Reprehenderit nam nemo esse provident accusamus nulla fuga dolorum aspernatur voluptatem, earum, explicabo repudiandae pariatur veniam, consequuntur a qui?

Lorem ipsum dolor sit amet consectetur adipisicing elit. Porro libero inventore voluptatibus ullam temporibus possimus ad odio fuga tempora atque qui consectetur eum voluptates numquam, commodi repellendus, architecto dolorum officiis nihil exercitationem, quidem impedit. Accusantium iusto vel placeat! Nihil ipsa aliquam maiores minus eum quos aperiam, minima aut nulla ipsam cumque, ex impedit assumenda velit ab tempore distinctio. Magnam repellendus doloribus suscipit itaque. Assumenda officia sed blanditiis modi iure quos voluptatibus labore? Dolorem nisi obcaecati adipisci officia eos porro laboriosam beatae ipsum molestiae eligendi recusandae libero sunt expedita veritatis dicta saepe vel, aliquam esse dolores voluptate, rerum, in enim mollitia deserunt? Ipsa omnis magni in eligendi vel facilis soluta vero officiis reprehenderit dolor eaque, cumque explicabo quo aliquam voluptatem distinctio vitae sit nam optio maxime nulla facere iste commodi hic. Quos repellendus sapiente accusamus quaerat, cumque facere perferendis quo iusto magnam omnis eveniet! Est sapiente laudantium quia, cum consectetur quo reiciendis assumenda, sunt, inventore quibusdam iusto dolores. Veritatis dolores blanditiis iusto omnis fugit cum magnam consectetur perspiciatis dolorem corporis sint sit ex esse nostrum, deserunt dolor delectus a voluptates amet eaque. Reprehenderit nam nemo esse provident accusamus nulla fuga dolorum aspernatur voluptatem, earum, explicabo repudiandae pariatur veniam, consequuntur a qui?

Lorem ipsum dolor sit amet consectetur adipisicing elit. Porro libero inventore voluptatibus ullam temporibus possimus ad odio fuga tempora atque qui consectetur eum voluptates numquam, commodi repellendus, architecto dolorum officiis nihil exercitationem, quidem impedit. Accusantium iusto vel placeat! Nihil ipsa aliquam maiores minus eum quos aperiam, minima aut nulla ipsam cumque, ex impedit assumenda velit ab tempore distinctio. Magnam repellendus doloribus suscipit itaque. Assumenda officia sed blanditiis modi iure quos voluptatibus labore? Dolorem nisi obcaecati adipisci officia eos porro laboriosam beatae ipsum molestiae eligendi recusandae libero sunt expedita veritatis dicta saepe vel, aliquam esse dolores voluptate, rerum, in enim mollitia deserunt? Ipsa omnis magni in eligendi vel facilis soluta vero officiis reprehenderit dolor eaque, cumque explicabo quo aliquam voluptatem distinctio vitae sit nam optio maxime nulla facere iste commodi hic. Quos repellendus sapiente accusamus quaerat, cumque facere perferendis quo iusto magnam omnis eveniet! Est sapiente laudantium quia, cum consectetur quo reiciendis assumenda, sunt, inventore quibusdam iusto dolores. Veritatis dolores blanditiis iusto omnis fugit cum magnam consectetur perspiciatis dolorem corporis sint sit ex esse nostrum, deserunt dolor delectus a voluptates amet eaque. Reprehenderit nam nemo esse provident accusamus nulla fuga dolorum aspernatur voluptatem, earum, explicabo repudiandae pariatur veniam, consequuntur a qui?

Lorem ipsum dolor sit amet consectetur adipisicing elit. Porro libero inventore voluptatibus ullam temporibus possimus ad odio fuga tempora atque qui consectetur eum voluptates numquam, commodi repellendus, architecto dolorum officiis nihil exercitationem, quidem impedit. Accusantium iusto vel placeat! Nihil ipsa aliquam maiores minus eum quos aperiam, minima aut nulla ipsam cumque, ex impedit assumenda velit ab tempore distinctio. Magnam repellendus doloribus suscipit itaque. Assumenda officia sed blanditiis modi iure quos voluptatibus labore? Dolorem nisi obcaecati adipisci officia eos porro laboriosam beatae ipsum molestiae eligendi recusandae libero sunt expedita veritatis dicta saepe vel, aliquam esse dolores voluptate, rerum, in enim mollitia deserunt? Ipsa omnis magni in eligendi vel facilis soluta vero officiis reprehenderit dolor eaque, cumque explicabo quo aliquam voluptatem distinctio vitae sit nam optio maxime nulla facere iste commodi hic. Quos repellendus sapiente accusamus quaerat, cumque facere perferendis quo iusto magnam omnis eveniet! Est sapiente laudantium quia, cum consectetur quo reiciendis assumenda, sunt, inventore quibusdam iusto dolores. Veritatis dolores blanditiis iusto omnis fugit cum magnam consectetur perspiciatis dolorem corporis sint sit ex esse nostrum, deserunt dolor delectus a voluptates amet eaque. Reprehenderit nam nemo esse provident accusamus nulla fuga dolorum aspernatur voluptatem, earum, explicabo repudiandae pariatur veniam, consequuntur a qui?

Lorem ipsum dolor sit amet consectetur adipisicing elit. Porro libero inventore voluptatibus ullam temporibus possimus ad odio fuga tempora atque qui consectetur eum voluptates numquam, commodi repellendus, architecto dolorum officiis nihil exercitationem, quidem impedit. Accusantium iusto vel placeat! Nihil ipsa aliquam maiores minus eum quos aperiam, minima aut nulla ipsam cumque, ex impedit assumenda velit ab tempore distinctio. Magnam repellendus doloribus suscipit itaque. Assumenda officia sed blanditiis modi iure quos voluptatibus labore? Dolorem nisi obcaecati adipisci officia eos porro laboriosam beatae ipsum molestiae eligendi recusandae libero sunt expedita veritatis dicta saepe vel, aliquam esse dolores voluptate, rerum, in enim mollitia deserunt? Ipsa omnis magni in eligendi vel facilis soluta vero officiis reprehenderit dolor eaque, cumque explicabo quo aliquam voluptatem distinctio vitae sit nam optio maxime nulla facere iste commodi hic.

</details>

ignatiusreza replied:
how did you came up with all these numbers? did you check existing templates?

kgrhartlage replied:
Suggestion based on what I felt is reasonable, plus some margin.

- step description serves no real purpose anyway, so 1000 should be more than enough.
- step titles should be much shorter than 200 chars, but I figured we don't want to limit users to much (despite the UI actually doing that)
- root descriptions is the only one that is maybe debatable, but 10.000 is a pretty long wall of text. Nothing in our workflow template repo come even anywhere near this

kgrhartlage commented on src/schemata/definitions/info.yml:
```diff
@@ -1,33 +1,35 @@
   title:
     type: string
-    title: Title
+    maxLength: 200
```
This field was added.

<details>
<summary>200 chars</summary>

Lorem ipsum dolor sit amet consectetur adipisicing elit. Porro libero inventore voluptatibus ullam temporibus possimus ad odio fuga tempora atque qui consectetur eum voluptates numquam, commodi repel.

</details>

kgrhartlage commented on src/schemata/definitions/step_template.yml:
```diff
@@ -1,22 +1,29 @@
+        properties:
+          title:
+            type: string
+            maxLength: 100
```
This field was added.

<details>
<summary>100 chars</summary>

Lorem ipsum dolor sit amet consectetur adipisicing elit. Porro libero inventore voluptatibus ullam.

</details>

ignatiusreza commented on src/schemata/definitions/step_template.yml:
```diff
@@ -1,22 +1,27 @@
+            type: string
+            description: |
+              A descriptive summary of the step.
       secrets:
         $ref: '#/definitions/secrets'
```
> I will rename this file from step_template to step if we agree on the other suggested changes. 

sure, go for it.. 

> I wasn't sure if GitHub would still pick up on the relevant changes, so I'm leaving this to a separate commit.

Git (not GitHub) do have some heuristic to detect file rename based on percentage of similarity of file content.. there's also a `git mv` command, but I'm not sure if it actually do anything different compared to standard `mv` command..

kgrhartlage commented:
I removed `slug` and `tags` from the root `info` schema, since slug does not work and tags don't do anything. I also added `additionalProperties: false` tot he root `info` schema. Yay or nay?

ignatiusreza commented:
Nay.. [slug is being used](https://github.com/labforward/laboperator-server/blob/develop/app/models/concerns/base_template.rb#L94)..

`tags`, should be okay to remove..

kgrhartlage commented:
> slug is being used

Does it work though? I remember trying it, but to no avail

ignatiusreza reviewed (dismissed):
LGTM, please proceed with the renaming..

ignatiusreza commented:
> Does it work though? I remember trying it, but to no avail

Yes and no.. the BE logic is working, the FE logic, not so much.. we definitely need to reassess if support for slug something that we want to support or not (across all resources), but we shouldn't just remove it somewhat randomly..

kgrhartlage commented:
I had to move the `step.yml` file also, since it otherwise messes up the pointers. An internal step can only exist within a workflow template anyway, so scoping this under `workflow_template` makes sense
